design-proposals/migration-policies.md

```diff
@@ -0,0 +1,224 @@
+the relevant label.
+
+## User Stories
+As an admin I would like to have different migration policies for different groups of VMs.
+* By "migration policies" I mean controlling one or more of the following migration aspects:
+    * Sensitivity to overall migration time
+    * Migration progress timeout
+    * Determining if and when post-copy / auto-converge are safe.
+    * Control number of migration threads, dirty rate (in the future)
+* By “different groups of VMs” I could mean:
+    * All VMs that have certain label defined
+    * All VMs that have certain label with certain value defined
+    * All VMs that belong to a namespace with certain label key/value
+* As an admin I know that some workloads are in higher priority than others and therefore need different 
+  migration policies.
+* As an admin I know that some workloads are not migration-friendly and demand careful configurations in order
+  for migrations to converge.
+* As an admin I want to have the ability to temporarily tune migrations for a specific event such as draining
+  a node or performing an upgrade.
+* As an admin I want that the migration policy would be as transparent as possible for the VM owner / creator.
+* As a user I want to be able to tell admins the workload characteristics of my VMs so that admins can create
+  policies and/or namespaces with the desired characteristics.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+The design will be similar to Kubernetes' [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+which is a similar concept. Therefore, a new MigrationPolicy CRD will be introduced. This CRD's spec will contain all
+relevant fields from KubevirtCR's [MigrationConfiguration](https://kubevirt.io/api-reference/master/definitions.html#_v1_migrationconfiguration)
+(i.e. which are not inherently cluster-wide configurations). In the future, additional
+configurations (or status related fields) that inherently relate to non-global context could be added,
+thus the configurations on NetworkPolicy and KubevirtCR's MigrationConfigurations may diverge to some degree.
+
+In addition, the new CRD's spec will include the following ways of specifying the groups of VMIs on which
+to apply the policy:
+* By VMI's labels
+* By namespace's labels
+
+All of these methods can be combined, for example a policy can require both VMI labels and namespace labels in
+order to match.
+
+## Policies' Precedence
+
+It is possible that a multiple policies apply to the same VMI. In such cases, the precedence is in the
+same order as the bullets above (VMI labels first, then namespace labels). It is not allowed to define
+two policies with the exact same selectors.
+
+If multiple policies apply to the same VMI:
+* The most detailed policy will be applied, that is, the policy with the highest number of matching labels
+* If multiple policies match to a VMI with the same number of matching labels, the policies will be sorted by the 
```

Comment:
this is fine, sorting by the migration policy names might be simpler though. Whatever makes the most sense to you. The big thing is the matching needs to be consistent.

Reply:
I agree that the only things matters here is being consistent, therefore I suggest we defer this decision to whatever is easiest to implement. If sorting by name is easier I don't see a reason not to use it instead :+1: